src/history/historyTracker.ts

```diff
@@ -427,14 +428,16 @@ export class HistoryTracker {
    * used to look like.
    */
   public addChange(cursorPosition = [new Position(0, 0)]): void {
-    const newText = this._getDocumentText();
+    // Get the current document's text which is conveniently equals to the zero based offset at the document's end.
+    const documentEnd = cursorPosition[0].getDocumentEnd();
+    const newTextLength = this.vimState.editor.document.offsetAt(documentEnd);
 
-    if (newText === this.oldText) {
+    // Check if the document's text changed.
+    if (newTextLength === this.oldText.length) {
```

Comment:
Perhaps we can just enumerate the special cases and deal with them separately, but doesn't this break with :s/abc/def, for instance?

Reply:
I believe you're referring to the case of doing `iabc<Esc>:s/abc/def\nu` which results in `|` when it should be `abc`. It does break that, sorry I hadn't considered that case and it seems we aren't currently testing for it. Can I add a test for it?

Reply:
Yeah a test for this would be great. If we go in this direction of optimization, I'd like to make sure we cover all the edge cases